utils/threadpool/prioritythreadpool.cpp

```diff
@@ -158,6 +179,14 @@ void PriorityThreadPool::threadFcn(const Priority preferredQueue) throw()
 
             if (jobQueues[queue].empty())
             {
+                // If this is an EXTRA thread due toother threads blocking, and all blockers are unblocked, 
+                // we don't want this one any more.
+                if (preferredQueue == EXTRA && stopExtra)
+                {
+                    extraThreads--;
```

Comment:
I am not clear on stopExtra semantics. It looks like this flag signals when there are no blocking threads. If so I can't get why don't we set extraThreads to zero instead of decrementing it?

Reply:
extraThreads counts how many extra threads are currently executing. When stopExtra is set, it effectively tells extra threads to go away when finished with whatever they're working on. When a thread goes away, it decrements extraThreads to reflect the current count of extra threads.

If there are blocked threads and extra threads, then this keeps a cap on how many extra threads will be created -- never more than the number of blocked threads.

However, as written, the number of extra threads may grow (if needed) as the number of blocked threads grows. The number of extra threads will not reduce until the number of blocked threads reaches 0 (thus setting stopExtra = true). I consider this to be OK, as blocked threads are a rarity. But if they do happen, they probably happen in bunches.

We may think to set extraThreads to zero in addJob() at the same time we set stopExtra to true. However, there will still be extra threads around until they finish whatever they were working on. During this period, extraThreads would be inaccurate. Meantime If threads block again, there could be confusion. stopExtra will be set to false and new extra threads will be created. There would be more extra threads in existence than extraThreads reflects. This would allow for more extraThreads than blocked threads. While not a killer, it's not what is intended.